### PR TITLE
add saving and loading models compatible with fsdp

### PR DESCRIPTION
**Pull request by kuba-krj**
# Description
todo

# Checklist
   - [x] Is this PR focused on a single feature without the possibility of being divided into smaller PRs?
   - [x] Are all variables set with logical defaults that are backward compatible?
   - [x] I attached link to neptune if it was necessary Link: [Neptune](https://app.neptune.ai/o/pmtest/org/llm-random/runs/table?viewId=9a15aa36-1063-49c4-807d-53393abb118a&detailsTab=monitoring&shortId=LLMRANDOM-7557&dash=charts&type=run&chartFilter=lr)
   - [x] Have you successfully executed the linter and tests?
   - [x] Are all functions concise and don't require splitting?
   - [x] Is there documentation for arguments and complex logic?
   - [x] Are there no temporary solutions in the code? Should a task be added to Trello to refine the code?
   - [x] Are all functions placed in appropriate files?
   - [x] Is the PR name meaningful and descriptive?
   - [x] Is PR description provided, or unnecessary?
  
Other questions:
   - [x] If I made changes to the requirements or anything related to the Singularity image I did generate new image and upload a new image to the clusters
   - [x] If this change is significant enough I notified all individuals working with this repository (@channel in slack channel)
   - [x] If there is a need for a second reviewer I requested additional review 

# Reviewer's checklist:
   - [ ] Does the Neptune link work and does it actually compare the situation before and after the PR?
   - [ ] Do I understand the code?
   - [ ] Do I think another reviewer is not needed or I commented that this PR need another reviewer?

### CODE REVIEW

**Comment by crewtool** [comment]
Why is the process of loading and saving model weights split into multiple parts across different files?  Could we have it in the same place?

File: research/conditional/train/cc_train.py [hunk 1]
```diff
@@ -86,15 +87,15 @@ def main(
...
-        ), "Do not add filename extensions (e.g. .pt or .pth) to save_weights_path! It is added automatically, along with step number."
-        random_string = generate_random_string(10)
-        args.save_weights_path = os.path.join(args.save_weights_path, random_string)
-        args.save_weights_path = os.path.abspath(args.save_weights_path)
-        os.makedirs(args.save_weights_path, exist_ok=True)
+
+    filename = args.save_weights_path.split("/")[-1]
+    assert (
+        "." not in filename
+    ), "Do not add filename extensions (e.g. .pt or .pth) to save_weights_path! It is added automatically, along with step number."
```
Comment (crewtool) [inline research/conditional/train/cc_train.py:94]: Move to check_args

File: lizrd/train/train_utils.py [hunk 2]
```diff
@@ -68,6 +70,16 @@ def get_model(
     ).to(last_gpu)
 
     model = llm.LLM(embedding_layer, encoder_tower, head)
+
+    if load_weights_path:
+        print(f"Loading weights...")
+        assert os.path.exists(
+            load_weights_path
+        ), f"Path {load_weights_path} does not exist"
```
Comment (crewtool) [inline lizrd/train/train_utils.py:78]: We can check it earlier i.e. before model creation

File: lizrd/train/train_utils.py [hunk 3]
```diff
@@ -68,6 +70,16 @@ def get_model(
...
+        checkpoint = torch.load(load_weights_path)
+        model.load_state_dict(checkpoint["model"], strict=False)
+        print(f"Loaded from {load_weights_path}")
```
Comment (crewtool) [inline lizrd/train/train_utils.py:81]: 1. Let wrap in into a function
2. Move it so it is next to `save_weights_path`

File: research/conditional/train/cc_train.py [hunk 4]
```diff
@@ -86,15 +87,15 @@ def main(
...
+    random_string = generate_random_string(10)
+    os.makedirs(args.save_weights_path, exist_ok=True)
+    args.save_weights_path = os.path.join(args.save_weights_path, random_string + ".pt")
+    args.save_weights_path = os.path.abspath(args.save_weights_path)
```
Comment (crewtool) [inline research/conditional/train/cc_train.py:98]: Is `os.path.abspath`  necessary?

File: research/conditional/train/cc_train.py [hunk 5]
```diff
@@ -86,15 +87,15 @@ def main(
...
-        args.save_weights_path = os.path.abspath(args.save_weights_path)
-        os.makedirs(args.save_weights_path, exist_ok=True)
+
+    filename = args.save_weights_path.split("/")[-1]
+    assert (
+        "." not in filename
+    ), "Do not add filename extensions (e.g. .pt or .pth) to save_weights_path! It is added automatically, along with step number."
+    random_string = generate_random_string(10)
+    os.makedirs(args.save_weights_path, exist_ok=True)
+    args.save_weights_path = os.path.join(args.save_weights_path, random_string + ".pt")
```
Comment (crewtool) [inline research/conditional/train/cc_train.py:97]: args should be prepared earlier i.e. before passing them to train_cc or we should have a function for preparing args for certain runner

Comment (crewtool) [inline research/conditional/train/cc_train.py:97]: maybe:
weights_filename  = f"{generate_random_string(10)}.pt"

File: research/conditional/train/cc_train.py [hunk 6]
```diff
@@ -225,6 +227,16 @@ def main(
...
+    if args.load_weights_path is not None:
+        checkpoint = torch.load(args.load_weights_path)
+        if args.fsdp_enabled:
+            full_osd = None
+            if rank == 0:
+                full_osd = checkpoint["optimizer"]
+            FSDP.scatter_full_optim_state_dict(full_osd, model)
+        else:
+            optimizer.load_state_dict(checkpoint["optimizer"])
+        step = checkpoint["step"]
```
Comment (crewtool) [inline research/conditional/train/cc_train.py:239]: make a function

**Review by crewtool — COMMENTED** [review]
Left some comments

File: research/conditional/utils/conditional_trainer.py [hunk 7]
```diff
@@ -449,28 +456,39 @@ def _save_weights(self, step):
...
-            checkpoint = torch.load(self.load_weights_path)
-            self.model.load_state_dict(checkpoint["model"], strict=False)
-            self.optimizer.load_state_dict(checkpoint["optimizer"])
-            if self.scaler is not None:
-                self.scaler.load_state_dict(checkpoint["scaler"])
-        else:
-            raise ValueError(
-                f"Path {self.load_weights_path} does not exist. Aborting ..."
-            )
+            print(f"Saving weights...")
```
[the remark this answers is not in the document]
Comment (kuba-krj) [reply research/conditional/utils/conditional_trainer.py:459]: fixed

File: research/conditional/train/cc_train.py (hunk 4 above)
Comment (kuba-krj) [reply research/conditional/train/cc_train.py:98]: Not sure, but it was here before

File: lizrd/train/train_utils.py (hunk 3 above)
Comment (kuba-krj) [reply lizrd/train/train_utils.py:81]: done

File: lizrd/train/train_utils.py (hunk 2 above)
Comment (kuba-krj) [reply lizrd/train/train_utils.py:78]: right, changed

File: research/conditional/train/cc_train.py (hunk 6 above)
Comment (kuba-krj) [reply research/conditional/train/cc_train.py:239]: done

File: research/conditional/train/cc_train.py (hunk 1 above)
Comment (kuba-krj) [reply research/conditional/train/cc_train.py:97]: done

File: research/conditional/train/cc_train.py (hunk 5 above)
Comment (kuba-krj) [reply research/conditional/train/cc_train.py:100]: done

File: lizrd/train/load_and_save_model.py [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+    StateDictType,
+)
+
+
+def get_checkpoint_from_path(load_weights_path: str) -> str:
+    print(f"Loading checkpoint from disk...")
+    assert os.path.exists(load_weights_path), f"Path {load_weights_path} does not exist"
+    checkpoint = torch.load(load_weights_path)
+    print(f"Loaded from {load_weights_path}")
+    return checkpoint
```
Comment (crewtool) [inline lizrd/train/load_and_save_model.py:17]:
```suggestion
    assert os.path.exists(load_weights_path), f"Path {load_weights_path} does not exist"
    print(f"Loading checkpoint from {load_weights_path}...")
    checkpoint = torch.load(load_weights_path)
    print(f"Checkpoint loaded")
    return checkpoint
```

File: research/conditional/train/cc_train.py [hunk 9]
```diff
@@ -86,15 +90,14 @@ def main(
...
-        ), "Do not add filename extensions (e.g. .pt or .pth) to save_weights_path! It is added automatically, along with step number."
-        random_string = generate_random_string(10)
-        args.save_weights_path = os.path.join(args.save_weights_path, random_string)
-        args.save_weights_path = os.path.abspath(args.save_weights_path)
+        weights_filename = f"{generate_random_string(10)}.pt"
         os.makedirs(args.save_weights_path, exist_ok=True)
+        save_weights_path = os.path.join(args.save_weights_path, weights_filename)
+        save_weights_path = os.path.abspath(args.save_weights_path)
+    else:
+        save_weights_path = None
```
Comment (crewtool) [inline research/conditional/train/cc_train.py:100]: Why not moving it to save_checkpoint?

**Review by crewtool — COMMENTED** [review]
Can we somehow test it automatically? 
Something like:
- set seed run steps 1-10
- set same seed run 1-5 then checkpoint
- set same seed, load checkpoint, run 6-10 and compare?

File: lizrd/train/load_and_save_model.py [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+
+
+def load_model_weights(
+    model: torch.nn.Module, checkpoint: dict[str, torch.Tensor]
+) -> torch.nn.Module:
+    print(f"Loading model weights...")
+    model.load_state_dict(checkpoint["model"], strict=False)
+    print(f"Loaded model weights")
+    return model
```
Comment (crewtool) [inline lizrd/train/load_and_save_model.py:26]: A bit missleading as model is already changed

File: lizrd/train/load_and_save_model.py [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+    print(f"Loading optimizer state...")
+    if isinstance(model, FSDP):
+        full_osd = None
+        if rank == 0:
+            full_osd = checkpoint["optimizer"]
+        FSDP.scatter_full_optim_state_dict(full_osd, model)
+    else:
+        optimizer.load_state_dict(checkpoint["optimizer"])
+    print(f"Loaded optimizer state")
+    return optimizer
```
Comment (crewtool) [inline lizrd/train/load_and_save_model.py:44]: same

File: lizrd/train/load_and_save_model.py [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+
+
+def load_scaler_state(
+    scaler: torch.cuda.amp.GradScaler,
+    checkpoint: dict[str, torch.Tensor],
+) -> torch.cuda.amp.GradScaler:
+    scaler.load_state_dict(checkpoint["scaler"])
+    return scaler
```
Comment (crewtool) [inline lizrd/train/load_and_save_model.py:52]: same

File: lizrd/train/load_and_save_model.py (hunk 11 above)
Comment (kuba-krj) [reply lizrd/train/load_and_save_model.py:44]: removed

File: research/conditional/train/cc_train.py (hunk 9 above)
Comment (kuba-krj) [reply research/conditional/train/cc_train.py:100]: moved

**Review by crewtool — APPROVED** [review]
Nicely done! LGTM!
